examples/step-57/step-57.cc

```diff
@@ -0,0 +1,878 @@
+    std::vector<Tensor<1, dim> >  phi_u                     (dofs_per_cell);
+    std::vector<Tensor<2, dim> >  grad_phi_u                (dofs_per_cell);
+    std::vector<double>           phi_p                     (dofs_per_cell);
+
+    typename DoFHandler<dim>::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+
+    for (; cell!=endc; ++cell)
+      {
+        fe_values.reinit(cell);
+
+        local_matrix = 0;
+        local_rhs    = 0;
+
+        fe_values[velocities].get_function_values(evaluation_point,
+                                                  present_velocity_values);
+
+        fe_values[velocities].get_function_gradients(evaluation_point,
+                                                     present_velocity_gradients);
+
+        fe_values[pressure].get_function_values(evaluation_point,
+                                                present_pressure_values);
+
+        // The assembly is similar to step-22. An additional term with gamma as the coefficient
+        // is the Lagrangian Argument(AL), which is assembled via grad-div stabilization.
+        // As we discussed in the introduction, the 1,1-block of system matrix should be
+        // zero. Since the pressure mass matrix is used while creating the preconditioner,
+        // we assemble it here and then move it into a separate SparseMatrix at the end (same as in step-22).
+        for (unsigned int q=0; q<n_q_points; ++q)
+          {
+            for (unsigned int k=0; k<dofs_per_cell; ++k)
+              {
+                div_phi_u[k]  =  fe_values[velocities].divergence (k, q);
+                grad_phi_u[k] =  fe_values[velocities].gradient(k, q);
+                phi_u[k]      =  fe_values[velocities].value(k, q);
+                phi_p[k]      =  fe_values[pressure]  .value(k, q);
+              }
+
+            for (unsigned int i=0; i<dofs_per_cell; ++i)
+              {
+                if (assemble_matrix)
+                  {
+                    for (unsigned int j=0; j<dofs_per_cell; ++j)
+                      {
+                        local_matrix(i, j) += (  viscosity*scalar_product(grad_phi_u[j], grad_phi_u[i])      //(gradU, gradV_u)
+                                                 + present_velocity_gradients[q]*phi_u[j]*phi_u[i]   //(d_U*grad_U_old, V_u)
+                                                 + grad_phi_u[j]*present_velocity_values[q]*phi_u[i] //(U_old*grad_d_U, V_u)
+                                                 - div_phi_u[i]*phi_p[j]                             //-(d_P, div_V_u)
+                                                 - phi_p[i]*div_phi_u[j]                             //-(div_d_U, V_p)
+                                                 + gamma*div_phi_u[j]*div_phi_u[i]                   //grad-div (div_u_j, div_u_i)
+                                                 + phi_p[i] * phi_p[j]) //pressure mass matrix
```

Comment:
Please indent all comment lines to the same position (or skip them, they are quite far inside the page margin already...)